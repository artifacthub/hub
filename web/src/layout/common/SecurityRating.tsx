import { isNull, isUndefined } from 'lodash';
import { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';

import { SecurityReportSummary, SeverityRating as SRating, VulnerabilitySeverity } from '../../types';
import { SEVERITY_RATING } from '../../utils/data';
import ElementWithTooltip from './ElementWithTooltip';
import Label from './Label';
import styles from './SecurityRating.module.css';

interface Props {
  summary?: SecurityReportSummary | null;
  className?: string;
  tooltipClassName?: string;
  onlyBadge: boolean;
  withLink?: string;
  withoutTooltip?: boolean;
  onlyIcon?: boolean;
  tooltipAlignment?: 'right' | 'left';
}

const SecurityRating = (props: Props) => {
  const navigate = useNavigate();
  const [severity, setSeverity] = useState<SRating | undefined>();

  useEffect(() => {
    if (props.summary) {
      let rating = SEVERITY_RATING.default!;
      for (const key in VulnerabilitySeverity) {
        const sev = key.toLowerCase();
        if (props.summary && props.summary.hasOwnProperty(sev) && (props.summary as any)[sev] > 0) {
          rating = SEVERITY_RATING[sev as VulnerabilitySeverity]!;
          break;
        }
      }
      setSeverity(rating);
    }
  }, [props.summary]);

  if (isNull(props.summary) || isUndefined(props.summary) || isUndefined(severity)) return null;

  const badge = (className?: string) => (
    <ElementWithTooltip
      className={className}
      element={
        props.onlyBadge ? (
          <small>
            <div
              className={`badge text-light fw-bold ${styles.badge} ${className}`}
              style={{
                backgroundColor: severity.color,
              }}
            >
              {severity.level}
            </div>
          </small>
        ) : (
          <Label
            className={`fw-bold ${className}`}
            text="Images Security Rating"
            textForSmallDevices="Security Rating"
            bgLeftIcon={severity.color}
            labelStyle="custom"
            icon={<span className={`position-relative ${styles.ratingLetter}`}>{severity.level}</span>}
            onlyIcon={props.onlyIcon}
          />
        )
      }
      tooltipWidth={300}
      tooltipClassName={`${styles.tooltip} ${props.tooltipClassName} ${props.onlyBadge ? styles.onlyBadgeTooltip : ''}`}
      tooltipMessage={
        <div className="d-flex flex-column">
          <div className="d-flex flex-row align-items-center my-1">
            <span
              className={`badge text-light fw-semibold me-2 ${styles.badge}`}
              style={{
                backgroundColor: SEVERITY_RATING.default!.color,
              }}
            >
              {SEVERITY_RATING.default!.level}
            </span>
            <div>No vulnerabilities found</div>
          </div>
          <div className="d-flex flex-row align-items-center my-1">
            <span
              className={`badge text-light fw-semibold me-2 ${styles.badge}`}
              style={{
                backgroundColor: SEVERITY_RATING[VulnerabilitySeverity.Low]!.color,
              }}
            >
              {SEVERITY_RATING[VulnerabilitySeverity.Low]!.level}
            </span>
            <div>
              Vulnerabilities of severity <span className="fw-bold">LOW</span> found
            </div>
          </div>
          <div className="d-flex flex-row align-items-center my-1">
            <span
              className={`badge text-light fw-semibold me-2 ${styles.badge}`}
              style={{
                backgroundColor: SEVERITY_RATING[VulnerabilitySeverity.Medium]!.color,
              }}
            >
              {SEVERITY_RATING[VulnerabilitySeverity.Medium]!.level}
            </span>
            <div>
              Vulnerabilities of severity <span className="fw-bold">MEDIUM</span> found
            </div>
          </div>
          <div className="d-flex flex-row align-items-center my-1">
            <span
              className={`badge text-light fw-semibold me-2 ${styles.badge}`}
              style={{
                backgroundColor: SEVERITY_RATING[VulnerabilitySeverity.High]!.color,
              }}
            >
              {SEVERITY_RATING[VulnerabilitySeverity.High]!.level}
            </span>
            <div>
              Vulnerabilities of severity <span className="fw-bold">HIGH</span> found
            </div>
          </div>
          <div className="d-flex flex-row align-items-center my-1">
            <span
              className={`badge text-light fw-semibold me-2 ${styles.badge}`}
              style={{
                backgroundColor: SEVERITY_RATING[VulnerabilitySeverity.Critical]!.color,
              }}
            >
              {SEVERITY_RATING[VulnerabilitySeverity.Critical]!.level}
            </span>
            <div>
              Vulnerabilities of severity <span className="fw-bold">CRITICAL</span> found
            </div>
          </div>
          <div className="d-flex flex-row align-items-center my-1">
            <span
              className={`badge text-light fw-semibold me-2 ${styles.badge}`}
              style={{
                backgroundColor: SEVERITY_RATING[VulnerabilitySeverity.UnKnown]!.color,
              }}
            >
              {SEVERITY_RATING[VulnerabilitySeverity.UnKnown]!.level}
            </span>
            <div>
              Vulnerabilities of severity <span className="fw-bold">UNKNOWN</span> found
            </div>
          </div>
        </div>
      }
      visibleTooltip={isUndefined(props.withoutTooltip) || !props.withoutTooltip}
      active
    />
  );

  return (
    <>
      {props.withLink ? (
        <div className={props.className}>
          <button
            className={`btn btn-link text-reset p-0 position-relative ${styles.link}`}
            onClick={(e) => {
              e.preventDefault();
              navigate({
                pathname: props.withLink,
                search: '?modal=security-report',
              });
            }}
            aria-label="Open security report"
            aria-hidden="true"
            tabIndex={-1}
          >
            {badge()}
          </button>
        </div>
      ) : (
        <>{badge(props.className)}</>
      )}
    </>
  );
};

export default SecurityRating;
