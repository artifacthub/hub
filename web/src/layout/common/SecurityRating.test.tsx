import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';

import SecurityRating from './SecutityRating';

const defaultProps = {
  summary: {
    low: 0,
    high: 7,
    medium: 1,
    unknown: 0,
    critical: 1,
  },
  onlyBadge: false,
};

describe('SecurityRating', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('creates snapshot', () => {
    const { asFragment } = render(<SecurityRating {...defaultProps} />);
    expect(asFragment).toMatchSnapshot();
  });

  it('renders label', async () => {
    render(<SecurityRating {...defaultProps} />);
    expect(screen.getByText('Images Security Rating')).toBeInTheDocument();
    expect(screen.getByText('F')).toBeInTheDocument();

    const badge = screen.getByTestId('elementWithTooltip');
    expect(badge).toBeInTheDocument();
    userEvent.hover(badge);

    expect(await screen.findByRole('tooltip')).toBeInTheDocument();
    expect(screen.getByText('No vulnerabilities found')).toBeInTheDocument();
    expect(screen.getAllByText(/Vulnerabilities of severity/g)).toHaveLength(5);
  });

  it('renders A label', () => {
    render(
      <SecurityRating
        summary={{
          low: 0,
          high: 0,
          medium: 0,
          unknown: 0,
          critical: 0,
        }}
        onlyBadge
      />
    );
    expect(screen.getByText('A')).toBeInTheDocument();
  });

  it('renders B label', () => {
    render(
      <SecurityRating
        summary={{
          low: 8,
          high: 0,
          medium: 0,
          unknown: 0,
          critical: 0,
        }}
        onlyBadge
      />
    );
    expect(screen.getByText('B')).toBeInTheDocument();
  });

  it('renders C label', () => {
    render(
      <SecurityRating
        summary={{
          low: 8,
          high: 0,
          medium: 3,
          unknown: 0,
          critical: 0,
        }}
        onlyBadge
      />
    );
    expect(screen.getByText('C')).toBeInTheDocument();
  });

  it('renders D label', () => {
    render(
      <SecurityRating
        summary={{
          low: 8,
          high: 9,
          medium: 3,
          unknown: 0,
          critical: 0,
        }}
        onlyBadge
      />
    );
    expect(screen.getByText('D')).toBeInTheDocument();
  });

  it('renders F label', () => {
    render(
      <SecurityRating
        summary={{
          low: 8,
          high: 9,
          medium: 3,
          unknown: 0,
          critical: 1,
        }}
        onlyBadge
      />
    );
    expect(screen.getByText('F')).toBeInTheDocument();
  });

  it('renders - label', () => {
    render(
      <SecurityRating
        summary={{
          low: 0,
          high: 0,
          medium: 0,
          unknown: 3,
          critical: 0,
        }}
        onlyBadge
      />
    );
    expect(screen.getByText('-')).toBeInTheDocument();
  });

  it('does not full label when onlyBadge is true', () => {
    render(<SecurityRating {...defaultProps} onlyBadge />);
    expect(screen.queryByText('Images Security Rating')).toBeNull();
  });

  it('does not render label', () => {
    const { container } = render(<SecurityRating summary={null} onlyBadge />);
    expect(container).toBeEmptyDOMElement();
  });
});
