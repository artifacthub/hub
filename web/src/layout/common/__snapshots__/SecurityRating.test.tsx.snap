// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SecurityRating creates snapshot 1`] = `[Function]`;
